File: src/test/deque/java/DequeTest.java
package test.deque.java;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;
import java.util.NoSuchElementException; 


public class DequeTest {

    Deque deque;
    @BeforeEach
    void setup() {
        deque = new Deque();
    }
    @Test
    public void testAppendLeft() {
        deque = new Deque(3);
        deque.appendLeft(10);
        assertEquals(10, deque.peekLeft());
        deque.appendLeft(20);
        assertEquals(20, deque.peekLeft());
        deque.appendLeft(3);
        assertEquals(3, deque.peekLeft());
        assertEquals(3,deque.getCapacity());
        deque.appendLeft(9);
        assertEquals(6,deque.getCapacity());

    }

    @Test
    public void testAppendRight() {
        
        deque = new Deque(3);
        deque.appendRight(10);
        assertEquals(10, deque.peekRight());
        deque.appendRight(20);
        assertEquals(20, deque.peekRight());
        deque.appendRight(3);
        assertEquals(3, deque.peekRight());
        assertEquals(3,deque.getCapacity());
        deque.appendRight(9);
        assertEquals(6,deque.getCapacity());

    }

    @Test
    public void testPopRight() {
        try {
            deque.popRight();
        } catch (NoSuchElementException e) {
            assertEquals("O Deque está vazio", e.getMessage());  
        }
        deque.appendLeft(10);
        deque.appendRight(7);
        deque.appendLeft(20);
        deque.appendRight(9);
        assertEquals(9, deque.popRight());
        assertEquals(7, deque.popRight());
        assertEquals(10, deque.popRight());
        assertEquals(20, deque.popRight());
    }

    @Test
    public void testPopLeft() {
        try {
            deque.popLeft();
        } catch (NoSuchElementException e) {
            assertEquals("O Deque está vazio", e.getMessage()); 
        }
        deque.appendLeft(10);
        deque.appendRight(7);
        deque.appendLeft(20);
        deque.appendRight(9);
        assertEquals(20, deque.popLeft());
        assertEquals(10, deque.popLeft());
        assertEquals(7, deque.popLeft());
        assertEquals(9, deque.popLeft());
    }

    @Test
    public void testPeekLeftAndRight() {
        try {
            deque.peekLeft();
        } catch (NoSuchElementException e) {
            assertEquals("O Deque está vazio", e.getMessage()); 
        }
        try {
            deque.peekLeft();
        } catch (NoSuchElementException e) {
            assertEquals("O Deque está vazio", e.getMessage()); 
        }
        deque.appendLeft(10);
        deque.appendLeft(20);
        deque.appendRight(7);
        deque.appendRight(9);
        assertEquals(20, deque.peekLeft());
        assertEquals(9, deque.peekRight());
        deque.popLeft();
        deque.popRight();
        assertEquals(10, deque.peekLeft());
        assertEquals(7, deque.popRight());

    }

    @Test
    public void testIsEmpty() {
        
        assertTrue(deque.isEmpty());
        deque.appendRight(10);
        assertFalse(deque.isEmpty());
        deque.popLeft();
        assertTrue(deque.isEmpty());
    }

    @Test
    public void testContains() {
        
        deque.appendRight(5);
        deque.appendRight(10);
        deque.appendRight(15);
        assertTrue(deque.contains(10));
        assertFalse(deque.contains(20));
        deque.popLeft();
        deque.popLeft();
        assertFalse(deque.contains(10));

    }
    
    @Test
    public void testResize() {
        Deque deque = new Deque(2);
        deque.appendRight(1);
        deque.appendRight(2);
        assertEquals(2, deque.getCapacity());        
        assertEquals(2, deque.getSize());
        deque.appendRight(3);
        assertEquals(4, deque.getCapacity());        
        assertEquals(3, deque.getSize());
        assertEquals(1, deque.popLeft());
        assertEquals(2, deque.popLeft());
        assertEquals(3, deque.popLeft());
        assertTrue(deque.isEmpty());
    }
    
}

File: src/test/deque/cpp/DequeTest.cpp
#include <iostream>
#include <cassert>
#include <stdexcept>
#include "../../deque/cpp/Deque.h"


Deque deque;

void setup() {
    deque = Deque(4);
}

void testAppendLeft() {
    setup();
    deque = Deque(3);
    deque.appendLeft(10);
    assert(deque.peekLeft() == 10);
    deque.appendLeft(20);
    assert(deque.peekLeft() == 20);
    deque.appendLeft(3);
    assert(deque.peekLeft() == 3);
    assert(deque.getCapacity() == 3);
    deque.appendLeft(9);
    assert(deque.getCapacity() == 6);
}

void testAppendRight() {
    setup();
    deque = Deque(3);
    deque.appendRight(10);
    assert(deque.peekRight() == 10);
    deque.appendRight(20);
    assert(deque.peekRight() == 20);
    deque.appendRight(3);
    assert(deque.peekRight() == 3);
    assert(deque.getCapacity() == 3);
    deque.appendRight(9);
    assert(deque.getCapacity() == 6);
}


void testPopLeft() {
     setup();
    try {
        deque.popLeft();
    } catch (const std::out_of_range& e) {
        assert(std::string(e.what()) == "O Deque está vazio"); 
    }
    deque.appendLeft(10);
    deque.appendRight(7);
    deque.appendLeft(20);
    deque.appendRight(9);
    assert(deque.popLeft() == 20);
    assert(deque.popLeft() == 10);
    assert(deque.popLeft() == 7);
    assert(deque.popLeft() == 9);
}

void testPopRight() {
    setup();
    try {
        deque.popRight();
    } catch (const std::out_of_range& e) {
        assert(std::string(e.what()) == "O Deque está vazio");
    }
    deque.appendLeft(10);
    deque.appendRight(7);
    deque.appendLeft(20);
    deque.appendRight(9);
    assert(deque.popRight() == 9);
    assert(deque.popRight() == 7);
    assert(deque.popRight() == 10);
    assert(deque.popRight() == 20);
}

void testPeekLeftAndRight() {
    setup();
    try {
        deque.peekLeft();
    } catch (const std::out_of_range& e) {
        assert(std::string(e.what()) == "O Deque está vazio");
    }
    try {
        deque.peekRight();
    } catch (const std::out_of_range& e) {
        assert(std::string(e.what()) == "O Deque está vazio");
    }
    deque.appendLeft(10);
    deque.appendLeft(20);
    deque.appendRight(7);
    deque.appendRight(9);
    assert(  deque.peekLeft() == 20);
    assert( deque.peekRight() == 9);
    deque.popLeft();
    deque.popRight();
    assert(deque.peekLeft() == 10);
    assert(deque.popRight() == 7);
}

void testIsEmpty() {
    setup();
    assert(deque.isEmpty());
    deque.appendRight(10);
    assert(!deque.isEmpty());
    deque.popLeft();
    assert(deque.isEmpty());
}

void testContains() {
    setup();
    deque.appendRight(5);
    deque.appendRight(10);
    deque.appendRight(15);
    assert(deque.contains(10));
    assert(!deque.contains(20));
    deque.popLeft();
    deque.popLeft();
    assert(!deque.contains(10));
}

void testResize() {
    setup();
    deque = Deque(2);
    deque.appendRight(1);
    deque.appendRight(2);
    assert(deque.getCapacity() == 2);
    assert(deque.getSize() == 2);
    deque.appendRight(3);
    assert(deque.getCapacity() == 4);
    assert(deque.getSize() == 3);
    deque.appendRight(4);
    assert(deque.getCapacity() == 4);
    deque.appendRight(5);
    assert(deque.getCapacity() == 8);
    deque.popLeft();
    deque.popLeft();
    deque.popLeft();
    assert(deque.isEmpty());
}

int main() {
    testAppendLeft();
    testAppendRight();
    testPopLeft();
    testPopRight();
    testPeekLeftAndRight();
    testIsEmpty();
    testContains();
    testResize();
    std::cout << "testes aprovados" << std::endl;
    return 0;
}
